### PR TITLE
Remove unused data and add fixes

### PR DESCRIPTION
**Pull request by MKlimenko**
This PR contains a set of fixes for the code including minor performance optimizations and removal of unused variables:

- Fix memory leak in [common/sampling.cpp]() by deleting the unused object before returning from a function 
- Replace single-character string with char comparison to improve performance in [examples/infill/infill.cpp]() and [examples/server/server.cpp]()
- Remove unused `std::vector` in [examples/llava/clip.cpp](), [pocs/vdot/vdot.cpp](), [tests/test-backend-ops.cpp], [examples/perplexity/perplexity.cpp]() and [tests/test-llama-grammar.cpp]() to get rid of extra dynamic allocations
- Replace expensive `strlen()` function call with a comparison to a null-terminator in [tests/get-model.cpp]()

If that sort of PRs is welcome, there is a bunch of warnings and potential memsize-related issues I can take care of.

### CODE REVIEW

File: tests/get-model.cpp [hunk 1]
```diff
@@ -11,7 +11,7 @@ char * get_model_or_exit(int argc, char *argv[]) {
 
     } else {
         model_path = getenv("LLAMACPP_TEST_MODELFILE");
-        if (!model_path || strlen(model_path) == 0) {
+        if (!model_path || model_path[0] == '\0') {
```
Comment (slaren) [inline tests/get-model.cpp:14]: In this case where performance is not a concern, I would prefer the clarity of `strlen`.

File: examples/perplexity/perplexity.cpp [hunk 2]
```diff
@@ -813,7 +813,6 @@ static void hellaswag_score(llama_context * ctx, const gpt_params & params) {
 
     llama_batch batch = llama_batch_init(n_ctx, 0, max_seq);
 
-    std::vector<float> tok_logits(n_vocab);
```
Comment (slaren) [inline examples/perplexity/perplexity.cpp:816]: This is not unused.

File: tests/test-backend-ops.cpp [hunk 3]
```diff
@@ -82,7 +82,6 @@ static std::vector<float> tensor_to_float(const ggml_tensor * t) {
 
     ggml_type_traits_t tt = ggml_internal_get_type_traits(t->type);
     size_t bs = ggml_blck_size(t->type);
-    std::vector<float> vq(ggml_blck_size(t->type));
```
Comment (slaren) [inline tests/test-backend-ops.cpp:85]: The one in the inner loop should be removed instead, this was intended to remove the allocations from the inner loop.

**Review by slaren — CHANGES_REQUESTED** [review]

File: tests/get-model.cpp (hunk 1 above)
Comment (cebtenzzre) [reply tests/get-model.cpp:14]: I would personally default to writing this as `!model_path || !*model_path`, but maybe that's too terse.

Comment (MKlimenko) [reply tests/get-model.cpp]: Yes, you're right. For single entry at the tests, I think it's okay to sacrifice a bit of performance for the sake of readability.

File: examples/perplexity/perplexity.cpp (hunk 2 above)
Comment (MKlimenko) [reply examples/perplexity/perplexity.cpp]: Yup, got a bit carried away with the static analyzer, reverted this

File: tests/test-backend-ops.cpp (hunk 3 above)
Comment (MKlimenko) [reply tests/test-backend-ops.cpp]: Got it, let's see if the current solution will pass the CI or if an additional cleanup for that vector is required

**Review by slaren — APPROVED** [review]

**Pull request merged**